rfcs/0010-txpool-dos-mitigation.md

```diff
@@ -0,0 +1,238 @@
+(send themselves $0) or resend them with a higher fee if they're processing too
+slow.
+
+We want to have this feature, but it allows an attacker to make proposers
+process transactions which won't eventually get mined (the ones that are
+replaced), which violates our first guiding principle. So we require the fee of
+the new transaction be at least `min_fee_increment` higher. This is the
+"standard" approach.
+
+#### Multiple pending transactions from the same account
+
+Since payments aren't instant, users may want to queue multiple outgoing
+transactions. This listed rules above cover this case, but things get
+complicated when you allow transaction replacement. Imagine Mallory broadcasts
+1000 valid transactions with sequential nonces, then replaces the first one with
+one that spends all the money in the account. The other 999 of them are now
+invalid and won't be mined, but the proposers still had to validate, store and
+gossip them, violating our first principle. So the new fee in the example needs
+to be at least `min_fee_increment` * 1000 higher.
+
+### When txpool size is exceeded: rules
+
+We have a set limit on txpool size: `max_txpool_size`, in transactions. If an
+incoming transaction would cause us to exceed the limit, we evict the lowest fee
+transaction from the mempool, or drop the incoming transaction if its fee is <=
+the lowest fee transaction in the mempool.
+
+If an incoming transaction has too low of a fee for us to accept, we count it as
+bad for the purposes of `bad_tx_blacklist_ratio`. Except in the case of
+replacement transactions. If we punished nodes for sending transactions without
+the replacement fee increment, an attacker could induce nodes to blacklist each
+other by sending them different transactions at the same nonce.
+
+### Constants
+
+-   `bad_tx_blacklist_ratio`: There are two situations where honest nodes send us
+    bad transactions. The first is when they are slightly behind us on the
+    consensus state, which is going to happen all the time just because of network
+    delay. The second is when they are substantially out of sync but haven't
+    noticed yet. After they notice they're supposed to disable gossip until
+    they're caught up. Dishonest nodes will send us invalid transactions in order
+    to waste our resources. Let's set this to 50% bad transactions. An attacker
+    then has to spend 2x as much resources to send a given amount of transactions.
+-   `good_faith_tx_credits`: Let's say we assume good faith if the first two
+    transactions we receive from a new peer are bad, but give up on them after
+    that. That means this value should be 2.
+-   `max_txpool_size`: This is interesting. If there were a fixed limit on the
+    number of transactions per block I'd say set it to an hour's worth or
+    something. But the limit is set by the parameters of the parallel scan, and
+    supposedly that'll become dynamic in the future, so I'm not sure. So let's say
+    1000?
```

Comment:
Bitcoin limits this by size in bytes right? https://jochen-hoenicke.de/queue/#0,24h Esp. since we'll have higher throughput, we may want to make `max_txpool_size` much larger.

Reply:
Bitcoin payments can have arbitrary sizes, but in Coda it's constant, so it should be strictly proportional. So there's not a real difference. Bytes is easier to think about in terms of resource consumption, but harder to implement in code.